File: examples/python/customer/handlers/state.py
"""Customer state management.

Contains state dataclasses and state rebuilding logic.
"""

from typing import Protocol

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains


class CustomerState(Protocol):
    """Protocol for customer state."""

    name: str
    email: str
    loyalty_points: int
    lifetime_points: int


class StateRebuildError(Exception):
    """Error during state rebuilding."""


def next_sequence(event_book: angzarr.EventBook | None) -> int:
    """Return the next event sequence number based on prior events."""
    if event_book is None or not event_book.pages:
        return 0
    return len(event_book.pages)


def rebuild_state(event_book: angzarr.EventBook | None) -> domains.CustomerState:
    """Rebuild customer state from events."""
    state = domains.CustomerState()

    if event_book is None or not event_book.pages:
        return state

    # Start from snapshot if present
    if event_book.snapshot and event_book.snapshot.state:
        state.ParseFromString(event_book.snapshot.state.value)

    # Apply events
    for page in event_book.pages:
        if not page.event:
            continue

        if page.event.type_url.endswith("CustomerCreated"):
            event = domains.CustomerCreated()
            page.event.Unpack(event)
            state.name = event.name
            state.email = event.email

        elif page.event.type_url.endswith("LoyaltyPointsAdded"):
            event = domains.LoyaltyPointsAdded()
            page.event.Unpack(event)
            state.loyalty_points = event.new_balance
            state.lifetime_points += event.points

        elif page.event.type_url.endswith("LoyaltyPointsRedeemed"):
            event = domains.LoyaltyPointsRedeemed()
            page.event.Unpack(event)
            state.loyalty_points = event.new_balance

    return state

File: examples/rust/customer/src/state.rs
//! Customer state management and reconstruction from events.

use common::proto::{CustomerCreated, CustomerState, LoyaltyPointsAdded, LoyaltyPointsRedeemed};
use common::{make_event_book, rebuild_from_events};
use prost::Message;

use angzarr::proto::{Cover, EventBook};

/// Protobuf type URL for CustomerState snapshots.
pub const STATE_TYPE_URL: &str = "type.examples/examples.CustomerState";

/// Rebuild customer state from an event book.
///
/// Applies events in order, starting from a snapshot if present.
pub fn rebuild_state(event_book: Option<&EventBook>) -> CustomerState {
    rebuild_from_events(event_book, apply_event)
}

pub fn apply_event(state: &mut CustomerState, event: &prost_types::Any) {
    if event.type_url.ends_with("CustomerCreated") {
        if let Ok(e) = CustomerCreated::decode(event.value.as_slice()) {
            state.name = e.name;
            state.email = e.email;
        }
    } else if event.type_url.ends_with("LoyaltyPointsAdded") {
        if let Ok(e) = LoyaltyPointsAdded::decode(event.value.as_slice()) {
            // Use facts (absolute values) for idempotent state reconstruction
            state.loyalty_points = e.new_balance;
            state.lifetime_points = e.new_lifetime_points;
        }
    } else if event.type_url.ends_with("LoyaltyPointsRedeemed") {
        if let Ok(e) = LoyaltyPointsRedeemed::decode(event.value.as_slice()) {
            state.loyalty_points = e.new_balance;
        }
    }
}

/// Apply an event and build an EventBook response with updated snapshot.
pub fn build_event_response(
    state: &CustomerState,
    cover: Option<Cover>,
    next_seq: u32,
    event_type_url: &str,
    event: impl Message,
) -> EventBook {
    let event_bytes = event.encode_to_vec();
    let any = prost_types::Any {
        type_url: event_type_url.to_string(),
        value: event_bytes.clone(),
    };
    let mut new_state = state.clone();
    apply_event(&mut new_state, &any);

    make_event_book(
        cover,
        next_seq,
        event_type_url,
        event_bytes,
        STATE_TYPE_URL,
        new_state.encode_to_vec(),
    )
}
